libdbus-sys/build_vendored.rs

```diff
@@ -0,0 +1,389 @@
+
+    // Add version definitions
+    let version_number = (DBUS_VERSION_MAJOR << 16) | (DBUS_VERSION_MINOR << 8) | DBUS_VERSION_MICRO;
+    config.set("VERSION", DBUS_VERSION);
+    config.set("DBUS_VERSION_STRING", DBUS_VERSION);
+    config.set("DBUS_VERSION", version_number);
+    config.set("DBUS_MAJOR_VERSION", DBUS_VERSION_MAJOR);
+    config.set("DBUS_MINOR_VERSION", DBUS_VERSION_MINOR);
+    config.set("DBUS_MICRO_VERSION", DBUS_VERSION_MICRO);
+
+    // Basic config (sync, disable asserts, enable checks)
+    config.set("DBUS_SESSION_BUS_CONNECT_ADDRESS", "\"autolaunch:\"");
+    config.enable("HAVE_DECL_MSG_NOSIGNAL");
+    config.enable("DBUS_USE_SYNC");
+    config.enable("DBUS_ENABLE_CHECKS");
+    config.disable("DBUS_DISABLE_CHECKS");
+    config.enable("DBUS_DISABLE_ASSERT");
+
+    // Target endian
+    match env::var("CARGO_CFG_TARGET_ENDIAN").as_deref() {
+        Ok("little") => {
+            config.disable("WORDS_BIGENDIAN");
+        },
+        Ok("big") => {
+            config.enable("WORDS_BIGENDIAN");
+        }
+        _ => unreachable!(),
+    }
+
+    // Platform dependent config
+    match env::var("CARGO_CFG_TARGET_OS").as_deref() {
+        Ok("linux") => {
+            config.enable("DBUS_UNIX");
+            config.enable("DBUS_HAVE_LINUX_EPOLL");
+            config.enable("HAVE_EPOLL");
+            config.enable("HAVE_ERRNO_H");
+            config.enable("HAVE_SOCKLEN_T");
+            config.enable("HAVE_GETPWNAM_R");
+            config.enable("HAVE_UNIX_FD_PASSING");
+            config.enable("HAVE_LOCALE_H");
+            config.enable("HAVE_DECL_ENVIRON");
+
+            // Disable
+            config.disable("HAVE_GETPEERUCRED");
+            config.disable("HAVE_CMSGCRED");
+
+            // System paths
+            let bus = format!("\"unix:path={}\"", DBUS_SYSTEM_BUS_DEFAULT_ADDRESS);
+            let dir = format!("\"{}\"", DBUS_SESSION_SOCKET_DIR);
+            let uid = format!("\"{}\"", DBUS_MACHINE_UUID_FILE);
+            config.set("DBUS_SESSION_SOCKET_DIR", dir);
+            config.set("DBUS_SYSTEM_BUS_DEFAULT_ADDRESS", bus);
+            config.set("DBUS_MACHINE_UUID_FILE", uid);
+
+            // GNU source required for struct ucred on linux
+            config.enable("_GNU_SOURCE");
+            cc.files(UNIX_SOURCES);
+        }
+        _ => return Err("Unsupported platform.".into()),
+    }
+    Ok(())
+}
+
+/// Vendored build entrypoint.
+///
+/// Instead of relying on cmake or meson. The build configuration was simple
+/// enough to port to cc directly. Allowing downstream crates to rely on the
+/// dbus crate without additional host dependencies.
+pub fn build_libdbus() -> Result<(), Box<dyn Error>> {
+    let mut compiler = cc::Build::new();
+    let mut config = Config::new();
+
+    // Ensure submodule is checked out
+    if !Path::new("vendor/dbus").exists() {
+        let _ = Command::new("git")
+            .args(&["submodule", "update", "--init", "vendor/dbus"])
```

Comment:
Hmm. Does this not mean that git the binary is now a build dependency of `dbus-rs`, if the vendored feature is enabled?

Reply:
If installing git is a step required for cross build then it should be documented.

Reply:
@diwic When published on `crates.io` it won't be required.  This is more of a development convenience for people who have cloned the repository without `--recursive`.

Using `libgit2-sys` as an example.

They have the C source as a git submodule here: https://github.com/rust-lang/git2-rs/tree/master/libgit2-sys
Then when published on crates.io it's archived with the full source at the specific commit: https://docs.rs/crate/libgit2-sys/0.14.0+1.5.0/source/libgit2/

The build.rs line is something they also have, but isn't explicitly required. When cargo pulls down the `libdbus-sys` source it will include the full `dbus` source as well. So the `git` line won't run since `vendor/dbus` exists.

Their build.rs: https://github.com/rust-lang/git2-rs/blob/master/libgit2-sys/build.rs#L27